net/http/authn/authn.go

```diff
@@ -72,8 +80,14 @@ func (a *API) Authenticate(req *http.Request) (*http.Request, error) {
 
 	// if there is no authentication at all, we return an "unauthenticated" error,
 	// which may be helpful when debugging
-	if len(X509Certs(ctx)) < 1 && err != nil {
-		return req, errors.New("unauthenticated")
+	if len(X509Certs(ctx)) < 1 && Token(ctx) == "" {
+		err := errors.New("unauthenticated")
+		if len(authnErrors) > 0 {
+			err = errors.WithDetailf(err, "Invalid credentials: %s", strings.Join(authnErrors, "; "))
```

Comment:
I think this (and two lines down) is a case where https://github.com/golang/go/wiki/CodeReviewComments#error-strings applies, and we don't want to capitalize the messages here.